### PR TITLE
Fix Labels and Points properties set

### PR DESCRIPTION
**Pull request by Czaki**
# Description

Fix properties of Labels and Points layers. Properly handling dict and Dataframe. I found it working on #2622

## Type of change

- [x] Bug-fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# References

# How has this been tested?

- [ ] example: the test suite for my feature covers cases x, y, and z
- [ ] example: all tests pass with my change

## Final checklist:
- [ ] My PR is the minimum possible work for the desired functionality
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] If I included new strings, I have used `trans.` to make them localizable.
      For more information see our [translations guide](https://napari.org/docs/dev/guides/translations.html).

### CODE REVIEW

File: napari/layers/labels/labels.py
Comment (Czaki) [inline napari/layers/labels/labels.py:481]: this is a small performance bugfix

**Comment by sofroniewn** [comment]
Thanks for the fix!! Code looks good here, but can you add tests too? It wasn't quite clear to me what was wrong before, so the test would help

**Comment by Czaki** [comment]
@kevinyamauchi I have a problem understanding the expected behavior of `_property_choices` of the `Points` class. If I good search it is only accessed inside the constructor. But still, in the constructor, its behavior is not constant. 

here
https://github.com/napari/napari/blob/3e35226f8f467fe957440635bfffb235e6d1d03e/napari/layers/points/points.py#L306-L308
pick unique elements

but here

https://github.com/napari/napari/blob/3e35226f8f467fe957440635bfffb235e6d1d03e/napari/layers/points/points.py#L310-L312
pick all elements. 

If I good understand there is the silent assumption that the user will provide a unique list of elements? 

The docstring for `properties` does not cover the case of empty data. 

Basing o the name it could be used for editing `properties` from GUI. 
But why this property is not updated in setter? 

Or how it should be updated in setter: overwritten, extended by missing values? Or something else? 

@sofroniewn suggests that you are the best person to ask.

**Review by sofroniewn — APPROVED** [review]
Thanks for the fix and adding the test here for labels. Maybe you can still add one for points. Otherwise this looks good to me! Maybe @kevinyamauchi can take a quick look before merge too

**Comment by Czaki** [comment]
@sofroniewn @jni @kevinyamauchi please review. 

As it was discussed at today meeting this PR should be the solution for the closest release and after this PR is finished the next PR with a general property engine will be stared

File: napari/layers/labels/labels.py
Comment (sofroniewn) [inline napari/layers/labels/labels.py]: Can you make this `Parameters` to match our other doc strings

**Review by sofroniewn — COMMENTED** [review]
> Fix properties of Labels and Points layers

Code changes look good, but can you explain in a little more detail what the above means. It's not clear to me what behavior was broken before this that is now fixed. Was there a regression from `0.4.7` or has this always been broken? It's a lot of changes close to the release, so I want to make sure we're all clear on what the expected changes are.

Maybe @jni who was in the dev meeting can also review this as they have more context.

File: napari/layers/labels/labels.py
Comment (Czaki) [reply napari/layers/labels/labels.py]: done

**Comment by Czaki** [comment]
> Code changes look good, but can you explain in a little more detail what the above means. It's not clear to me what behavior was broken before this that is now fixed. Was there a regression from `0.4.7` or has this always been broken? It's a lot of changes close to the release, so I want to make sure we're all clear on what the expected changes are.

It looks like it always was broken. My final target is #2658 and I creat this PR because I found that setting the properties was broken.

**Comment by sofroniewn** [comment]
> It looks like it always was broken. My final target is #2658 and I creat this PR because I found that setting the properties was broken.

Ok, nice, let's hold this till 0.4.9 then and give a few more people a chance to review

File: napari/layers/labels/labels.py
Comment (jni) [inline napari/layers/labels/labels.py]:
```suggestion
        """Convert an input properties value to a standard dict-of-columns format.

        Parameters
        ----------
```

Comment (jni) [reply napari/layers/labels/labels.py]: Probably this should be a function rather than a method, but it is not urgent to fix this in this PR.

File: napari/layers/points/points.py
Comment (jni) [inline napari/layers/points/points.py]:
```suggestion
                    "Property choices should be passed as property_choices, not properties. This warning will become an error in version 0.4.11.",
                ),
```

Translated strings need to be on one line until we improve our translation engine.

File: napari/layers/points/points.py
Comment (jni) [inline napari/layers/points/points.py]: Can you make whatever `True` is here a keyword-only argument? This signature is not possible to interpret without going to the definition.

File: napari/layers/points/points.py
Comment (jni) [inline napari/layers/points/points.py:383]: Is black really happy with this??? I am surprised it doesn't need to be indented...

File: napari/layers/points/points.py
Comment (jni) [inline napari/layers/points/points.py]:
```suggestion
        """Return properties in a normalized dict-of-columns format.

        Parameters
        ----------
        properties : Union[dict, DataFrame]
            properties to be transformed
        property_choices : Dict[str, np.ndarray]
            previous choices
        save_choices : bool
            preserve property choices that are not available in input columns.

        Returns
        -------
        properties (dict):
            properties dictionary
```

File: napari/layers/points/points.py
Comment (jni) [inline napari/layers/points/points.py]: I think the previous approach is more readable. Plus, relying on implicit bool conversion can often yield unexpected results.

File: napari/layers/points/points.py
Comment (jni) [inline napari/layers/points/points.py]: Generally we prefer to have a single return statement for a function, or have return statements really close together.

File: napari/layers/points/points.py
Comment (jni) [inline napari/layers/points/points.py]: As above — it is better not to rely on implicit bool conversion.

File: napari/layers/utils/layer_utils.py
Comment (jni) [inline napari/layers/utils/layer_utils.py]: Thanks for spotting this. Could you update the docstring to match? :pray:

**Review by jni — APPROVED** [review]
Thank you @Czaki! Apologies that we took so long to review this! I have left mostly minor comments, could you please address them and we can get this in? It'll be great to unify all of our properties handling after this!

File: napari/layers/labels/labels.py
Comment (Czaki) [reply napari/layers/labels/labels.py]: this call other classmethod/staticmethod `cls._validate_properties` so need to be classmethod

File: napari/layers/points/points.py
Comment (Czaki) [reply napari/layers/points/points.py:383]: this is black output. Before black it was one line.

File: napari/layers/points/points.py
Comment (Czaki) [reply napari/layers/points/points.py]: Nine lives above it is explicity created as set `self._selected_data = set(selected_data)`.
In general bool conversion in a well-written container should be no slower than `len` and in some cases should be faster. Because of this, it is suggested to use `if not self._selected_data` rather than `if len(self._selected_data) == 0:` but I'm not tied with this change.

**Comment by jni** [comment]
Thanks for addressing review comments @Czaki! @sofroniewn this is ready to be merged as far as I'm concerned. 🎉

**Pull request merged**
